### ISSUE TITLE
Nodes closed in TicketReview may still be active in AM

### ISSUE THREAD

**Issue by hinchliff** [RENCI-NRIG/orca5#123]
After TicketReview closes some reservations, e.g.

```
Node name: Workers-1
Reservation ID: 529eb8a0-d857-4b52-b65c-9268b61af6e5
Node reservation state: Closed
Reservation notice: Reservation 529eb8a0-d857-4b52-b65c-9268b61af6e5 (Slice 106_hincha) is in state [Closed,None], err=TicketReview: Closing reservation due to failure in Slice.
```

And after deleting the Slice, the AM may still think those reservations are active:

```
pequod:show>show reservations for all actor gwu-vm-am state active
529eb8a0-d857-4b52-b65c-9268b61af6e5    gwu-vm-am
    Slice: 1519ca6e-3a35-4de1-9615-a7a8aaa4b8e2
    1    gwuvmsite.vm    [ active, nascent]    
    Notices: Reservation 529eb8a0-d857-4b52-b65c-9268b61af6e5 (Slice 106_hincha) is in state [Active,None]
    Start: Thu Apr 20 10:08:53 EDT 2017    End: Fri Apr 21 10:08:54 EDT 2017    Requested end: Fri Apr 21 10:08:54 EDT 2017
79baeb68-388f-48fa-b343-7a0e27356137    gwu-vm-am
    Slice: 1519ca6e-3a35-4de1-9615-a7a8aaa4b8e2
    1    gwuvmsite.vm    [ active, nascent]    
    Notices: Reservation 79baeb68-388f-48fa-b343-7a0e27356137 (Slice 106_hincha) is in state [Active,None]
    Start: Thu Apr 20 10:08:53 EDT 2017    End: Fri Apr 21 10:08:54 EDT 2017    Requested end: Fri Apr 21 10:08:54 EDT 2017

Total: 2 reservations
```

It's possible that this only happens when TicketReview closes some reservations _after_ other parts of the Slice have already become active.   In this case, I was testing a Multipoint slice.  Two of the VLANs had become active, and then on of the VMs failed, which caused TicketReview to close the VM included in this message.

**Comment by YufengXin** [comment]
Detailed information from my slice: 1 broadcast link connecting 3 VMs, all intra-rack, one VM failed due to wrong image information.
(1) SM: all 3 VMs were closed, but the vlan reservation stays active. The slice is alive
(2) VM AM: all the 2 active VMs keep alive without being closed
(3) Then I closed these 2 VM reservation at the AM from pequod
(4) Then go back to Flukes to close the slice. From pequod, the VLAN reservation stays at "close wait" state (in SM). 	
1	rcivmsite.vlan	[ closewait, nascent]	
in AM, the reservation seems to be gone?

**Comment by YufengXin** [comment]
Actually, if the failed VM failed in a later stage, the other 2 VMs would not be closed in SM.

**Comment by mcevik0** [comment]
While working on RADICS slices created by AHAB and connecting ~10 sites, Paul encountered some problems primarily caused by leaked OSCARS VLANs. I think this is related with this ORCA issue. Based on my observations, I wanted to give some information. This can be better observed in big slices, some parts of the slice become active (eg OSCARS circuits), then a reservation fails and all reservations transition to "closed" state. However, OSCARS circuits remain active, probably handler needs to be invoked properly. I saw the reservations were closed on exo-sm, but I did not check ion-net-site at that time, probably they were closed on ion-net-site, too. Because, all I did was cancelling the OSCARS circuits through OSCARS portal after verifying no active ion.vlan reservations existed on exo-sm and new slices could be created properly without OSCARS failures for already occupied VLAN tags.

**Comment by ibaldin** [comment]
per Anirban @anriban via Slack
```
Just wanted to add my two cents to this.. I believe it is just not about leaking circuits.. I saw a clear case today of a leaking VM.. Story in short: A VM in an inter-domain slice fails for unknown reason -> all VMs and ion vlans go to  closed state but the net vlans remain in Active state -> interestingly, a lone VM not connected to other elements of the slice remain in Active state. The successful VMs that went to Closed state remain in OpenStack because the leave handler is never invoked for those VMs; hence declared being leaked..
```

**Comment by hinchliff** [comment]
Adding some pictures of what I think is happening, in case we want to look at examples during the meeting.

This picture shows what TicketReview was designed for:
![ticketreview_ok](https://cloud.githubusercontent.com/assets/10090677/25957882/4b2e7fb4-363d-11e7-92a3-e8f96a7dea96.png)

All of the reservations are still in the SM.  When reservation `1` is failed _by the SM_, TicketReview should Close reservations `2` and `3`.   I think this part is working correctly.

TicketReview was designed to prevent any of the reservations from crossing that blue dashed line from SM to AM until all of the reservations had been ticketed.

However this picture shows where TicketReview is currently acting incorrectly:
![ticketreview_fix](https://cloud.githubusercontent.com/assets/10090677/25959496/c3ff7afc-3641-11e7-9da2-05a72af574bc.png)

Reservation `2` has failed at the AM.  Reservation `3` has been redeemed, but is not yet Active (or Failed).  TicketReview is called on the final reservation, `4`, still waiting to be redeemed, and TicketReview is incorrectly closing both reservation `3` and `4`.

The fix I added in e5405c4ea58e3e2c301522eaab783be289b694b0 should prevent any reservations like `3` from being closed (after the AM has already seen them), by checking whether they having a `pending` state of `redeeming`.  (In this fix, neither `3` nor `4` would get closed by TicketReview.)

If reservation `3` was not present in this scenario, reservation `4` might get closed by TicketReview.  But since it had not yet been passed to the AM, it would not result in a leak.
